In [1]:
## Dataset Airbnb

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("listings.csv")
df

,id,nombre,id_anfitrion,nombre_anfitrion,grupo_colonia,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,license
0,35797,Villa Dante,153786,Dici,NaN,Cuajimalpa de Morelos,19.382830,-99.271780,Entire home/apt,3673.0,1,0,NaN,NaN,1,363,0,NaN
1,44616,Condesa Haus,196253,Fernando,NaN,Cuauhtémoc,19.411620,-99.177940,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1,NaN
2,56074,Great space in historical San Rafael,265650,Maris,NaN,Cuauhtémoc,19.439770,-99.156050,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1,NaN
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,NaN,Cuauhtémoc,19.411520,-99.168570,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1,NaN
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,NaN,Coyoacán,19.354480,-99.162170,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
27046,1518356968266486778,Suite Frida,674644167,Heyliana,NaN,Venustiano Carranza,19.442240,-99.113440,Private room,1096.0,1,0,NaN,NaN,5,365,0,NaN
27047,1518388105556721163,Habitación a un costado de Ciudad Universitaria,222854336,Marco Antonio,NaN,Coyoacán,19.308017,-99.168158,Private room,260.0,1,0,NaN,NaN,1,365,0,NaN
27048,1518433354720281854,Espacioso departamento en Anzures,470800231,Hector Fernando,NaN,Miguel Hidalgo,19.434460,-99.174010,Entire home/apt,996.0,1,0,NaN,NaN,20,365,0,NaN
27049,1518462568893975567,Casa Roma Baja 05,710857718,Camila,NaN,Cuauhtémoc,19.406435,-99.160934,Entire home/apt,458.0,1,0,NaN,NaN,7,363,0,NaN


In [4]:
df.shape

(27051, 18)

In [5]:
df.columns

Index(['id', 'nombre', 'id_anfitrion', 'nombre_anfitrion', 'grupo_colonia',
       'colonia', 'latitud', 'longitud', 'tipo_habitacion', 'precio',
       'noches_minimas', 'numero_resenas', 'ultima_resena', 'resenas_por_mes',
       'propiedades_anfitrion', 'disponibilidad_365', 'number_of_reviews_ltm',
       'license'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27051 entries, 0 to 27050
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     27051 non-null  int64  
 1   nombre                 27051 non-null  object 
 2   id_anfitrion           27051 non-null  int64  
 3   nombre_anfitrion       27045 non-null  object 
 4   grupo_colonia          0 non-null      float64
 5   colonia                27051 non-null  object 
 6   latitud                27051 non-null  float64
 7   longitud               27051 non-null  float64
 8   tipo_habitacion        27051 non-null  object 
 9   precio                 23567 non-null  float64
 10  noches_minimas         27051 non-null  int64  
 11  numero_resenas         27051 non-null  int64  
 12  ultima_resena          23650 non-null  object 
 13  resenas_por_mes        23650 non-null  float64
 14  propiedades_anfitrion  27051 non-null  int64  
 15  di

In [7]:
df.describe()

,id,id_anfitrion,grupo_colonia,latitud,longitud,precio,noches_minimas,numero_resenas,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,license
count,2.705100e+04,2.705100e+04,0.0,27051.000000,27051.000000,23567.000000,27051.000000,27051.000000,23650.000000,27051.000000,27051.000000,27051.000000,0.0
mean,7.003556e+17,2.429439e+08,NaN,19.405490,-99.165385,1792.540841,4.580866,53.777679,1.854855,14.559868,232.686333,15.691028,NaN
std,5.702714e+17,2.062579e+08,NaN,0.042390,0.033535,13230.940558,24.784605,85.143043,2.281154,31.892178,122.270858,23.615517,NaN
min,3.579700e+04,7.365000e+03,NaN,19.177848,-99.339630,61.000000,1.000000,0.000000,0.010000,1.000000,0.000000,0.000000,NaN
25%,4.410472e+07,5.621443e+07,NaN,19.392489,-99.178499,643.000000,1.000000,4.000000,0.450000,1.000000,140.000000,0.000000,NaN
50%,8.307368e+17,1.766835e+08,NaN,19.415390,-99.167100,1039.000000,2.000000,21.000000,1.260000,3.000000,269.000000,7.000000,NaN
75%,1.217382e+18,4.288830e+08,NaN,19.432015,-99.153731,1611.000000,2.000000,69.000000,2.620000,11.000000,341.000000,23.000000,NaN
max,1.518561e+18,7.208445e+08,NaN,19.561010,-98.963360,900000.000000,1125.000000,1434.000000,119.460000,221.000000,365.000000,665.000000,NaN


In [8]:
df.tail()

,id,nombre,id_anfitrion,nombre_anfitrion,grupo_colonia,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,license
27046,1518356968266486778,Suite Frida,674644167,Heyliana,NaN,Venustiano Carranza,19.442240,-99.113440,Private room,1096.0,1,0,NaN,NaN,5,365,0,NaN
27047,1518388105556721163,Habitación a un costado de Ciudad Universitaria,222854336,Marco Antonio,NaN,Coyoacán,19.308017,-99.168158,Private room,260.0,1,0,NaN,NaN,1,365,0,NaN
27048,1518433354720281854,Espacioso departamento en Anzures,470800231,Hector Fernando,NaN,Miguel Hidalgo,19.434460,-99.174010,Entire home/apt,996.0,1,0,NaN,NaN,20,365,0,NaN
27049,1518462568893975567,Casa Roma Baja 05,710857718,Camila,NaN,Cuauhtémoc,19.406435,-99.160934,Entire home/apt,458.0,1,0,NaN,NaN,7,363,0,NaN
27050,1518561444972990747,Departamento con Terraza privada Roma Norte,362149590,Ilse,NaN,Cuauhtémoc,19.413682,-99.154815,Entire home/apt,3082.0,1,0,NaN,NaN,4,365,0,NaN


In [9]:
df.head()

,id,nombre,id_anfitrion,nombre_anfitrion,grupo_colonia,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,license
0,35797,Villa Dante,153786,Dici,NaN,Cuajimalpa de Morelos,19.38283,-99.27178,Entire home/apt,3673.0,1,0,NaN,NaN,1,363,0,NaN
1,44616,Condesa Haus,196253,Fernando,NaN,Cuauhtémoc,19.41162,-99.17794,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1,NaN
2,56074,Great space in historical San Rafael,265650,Maris,NaN,Cuauhtémoc,19.43977,-99.15605,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1,NaN
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,NaN,Cuauhtémoc,19.41152,-99.16857,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1,NaN
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,NaN,Coyoacán,19.35448,-99.16217,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8,NaN


In [10]:
valores_faltantes =  df.isna().sum().sort_values(ascending= False)
valores_faltantes

license                  27051
grupo_colonia            27051
precio                    3484
resenas_por_mes           3401
ultima_resena             3401
nombre_anfitrion             6
nombre                       0
id                           0
colonia                      0
id_anfitrion                 0
tipo_habitacion              0
longitud                     0
latitud                      0
numero_resenas               0
noches_minimas               0
propiedades_anfitrion        0
disponibilidad_365           0
number_of_reviews_ltm        0
dtype: int64

In [11]:
# Eliminacion de columna que no contiene ningún dato

del df["license"]  # Es una función de Python
# df.drop(columns=['license'])  -- Mismo resultado pero usando pandas

In [12]:
df

,id,nombre,id_anfitrion,nombre_anfitrion,grupo_colonia,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
0,35797,Villa Dante,153786,Dici,NaN,Cuajimalpa de Morelos,19.382830,-99.271780,Entire home/apt,3673.0,1,0,NaN,NaN,1,363,0
1,44616,Condesa Haus,196253,Fernando,NaN,Cuauhtémoc,19.411620,-99.177940,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1
2,56074,Great space in historical San Rafael,265650,Maris,NaN,Cuauhtémoc,19.439770,-99.156050,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,NaN,Cuauhtémoc,19.411520,-99.168570,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,NaN,Coyoacán,19.354480,-99.162170,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
27046,1518356968266486778,Suite Frida,674644167,Heyliana,NaN,Venustiano Carranza,19.442240,-99.113440,Private room,1096.0,1,0,NaN,NaN,5,365,0
27047,1518388105556721163,Habitación a un costado de Ciudad Universitaria,222854336,Marco Antonio,NaN,Coyoacán,19.308017,-99.168158,Private room,260.0,1,0,NaN,NaN,1,365,0
27048,1518433354720281854,Espacioso departamento en Anzures,470800231,Hector Fernando,NaN,Miguel Hidalgo,19.434460,-99.174010,Entire home/apt,996.0,1,0,NaN,NaN,20,365,0
27049,1518462568893975567,Casa Roma Baja 05,710857718,Camila,NaN,Cuauhtémoc,19.406435,-99.160934,Entire home/apt,458.0,1,0,NaN,NaN,7,363,0


In [13]:
df_copia =  df.copy()
df_copia.head()

,id,nombre,id_anfitrion,nombre_anfitrion,grupo_colonia,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
0,35797,Villa Dante,153786,Dici,NaN,Cuajimalpa de Morelos,19.38283,-99.27178,Entire home/apt,3673.0,1,0,NaN,NaN,1,363,0
1,44616,Condesa Haus,196253,Fernando,NaN,Cuauhtémoc,19.41162,-99.17794,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1
2,56074,Great space in historical San Rafael,265650,Maris,NaN,Cuauhtémoc,19.43977,-99.15605,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,NaN,Cuauhtémoc,19.41152,-99.16857,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,NaN,Coyoacán,19.35448,-99.16217,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8


In [14]:
df_copia = df_copia.drop(columns = ['grupo_colonia'])

In [15]:
df_copia.head()

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
0,35797,Villa Dante,153786,Dici,Cuajimalpa de Morelos,19.38283,-99.27178,Entire home/apt,3673.0,1,0,NaN,NaN,1,363,0
1,44616,Condesa Haus,196253,Fernando,Cuauhtémoc,19.41162,-99.17794,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1
2,56074,Great space in historical San Rafael,265650,Maris,Cuauhtémoc,19.43977,-99.15605,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,Cuauhtémoc,19.41152,-99.16857,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,Coyoacán,19.35448,-99.16217,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8


In [16]:
df_copia.columns

Index(['id', 'nombre', 'id_anfitrion', 'nombre_anfitrion', 'colonia',
       'latitud', 'longitud', 'tipo_habitacion', 'precio', 'noches_minimas',
       'numero_resenas', 'ultima_resena', 'resenas_por_mes',
       'propiedades_anfitrion', 'disponibilidad_365', 'number_of_reviews_ltm'],
      dtype='object')

In [17]:
df_copia.isna().mean().sort_values(ascending= False) * 100


precio                   12.879376
ultima_resena            12.572548
resenas_por_mes          12.572548
nombre_anfitrion          0.022180
colonia                   0.000000
id                        0.000000
nombre                    0.000000
id_anfitrion              0.000000
tipo_habitacion           0.000000
longitud                  0.000000
latitud                   0.000000
noches_minimas            0.000000
numero_resenas            0.000000
propiedades_anfitrion     0.000000
disponibilidad_365        0.000000
number_of_reviews_ltm     0.000000
dtype: float64

In [18]:
## Ver cuantos tipos de habitación tenemos

df_copia['tipo_habitacion'].unique()

array(['Entire home/apt', 'Private room', 'Hotel room', 'Shared room'],
      dtype=object)

In [19]:
df_copia['tipo_habitacion'].value_counts()
df_copia['tipo_habitacion'].value_counts(normalize = True) # para conocer porcentaje

tipo_habitacion
Entire home/apt    0.654800
Private room       0.332520
Shared room        0.009501
Hotel room         0.003179
Name: proportion, dtype: float64

In [20]:
df_copia['colonia'].value_counts()

colonia
Cuauhtémoc                12514
Miguel Hidalgo             4573
Benito Juárez              3087
Coyoacán                   1727
Álvaro Obregón              978
Venustiano Carranza         907
Tlalpan                     806
Gustavo A. Madero           515
Iztacalco                   482
Cuajimalpa de Morelos       410
Azcapotzalco                387
Iztapalapa                  302
Xochimilco                  155
La Magdalena Contreras      138
Tláhuac                      40
Milpa Alta                   30
Name: count, dtype: int64

In [21]:
##Número de reseña

In [22]:
df_copia['numero_resenas'].max()

1434

In [23]:
df_copia['numero_resenas'].min()

0

In [24]:

df_copia['ultima_resena'] = pd.to_datetime(df_copia['ultima_resena'] )
type(df_copia['ultima_resena'] )
df_copia.columns

Index(['id', 'nombre', 'id_anfitrion', 'nombre_anfitrion', 'colonia',
       'latitud', 'longitud', 'tipo_habitacion', 'precio', 'noches_minimas',
       'numero_resenas', 'ultima_resena', 'resenas_por_mes',
       'propiedades_anfitrion', 'disponibilidad_365', 'number_of_reviews_ltm'],
      dtype='object')

In [25]:
df_copia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27051 entries, 0 to 27050
Data columns (total 16 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     27051 non-null  int64         
 1   nombre                 27051 non-null  object        
 2   id_anfitrion           27051 non-null  int64         
 3   nombre_anfitrion       27045 non-null  object        
 4   colonia                27051 non-null  object        
 5   latitud                27051 non-null  float64       
 6   longitud               27051 non-null  float64       
 7   tipo_habitacion        27051 non-null  object        
 8   precio                 23567 non-null  float64       
 9   noches_minimas         27051 non-null  int64         
 10  numero_resenas         27051 non-null  int64         
 11  ultima_resena          23650 non-null  datetime64[ns]
 12  resenas_por_mes        23650 non-null  float64       
 13  p

In [26]:
# Fecha de la primer reseña
df_copia['ultima_resena'].min()

Timestamp('2013-12-21 00:00:00')

In [27]:
##Ultima reseña
df_copia['ultima_resena'].max()

Timestamp('2025-09-28 00:00:00')

In [28]:
# La colonia con más propiedades en Airbnb
df_copia['colonia'].value_counts()

colonia
Cuauhtémoc                12514
Miguel Hidalgo             4573
Benito Juárez              3087
Coyoacán                   1727
Álvaro Obregón              978
Venustiano Carranza         907
Tlalpan                     806
Gustavo A. Madero           515
Iztacalco                   482
Cuajimalpa de Morelos       410
Azcapotzalco                387
Iztapalapa                  302
Xochimilco                  155
La Magdalena Contreras      138
Tláhuac                      40
Milpa Alta                   30
Name: count, dtype: int64

In [29]:
# Cantidad de colonias en total
len(df_copia["colonia"].unique())
df_copia["colonia"].unique()

array(['Cuajimalpa de Morelos', 'Cuauhtémoc', 'Coyoacán',
       'Miguel Hidalgo', 'Benito Juárez', 'Iztacalco', 'Tlalpan',
       'Venustiano Carranza', 'Gustavo A. Madero', 'Xochimilco',
       'Álvaro Obregón', 'Iztapalapa', 'La Magdalena Contreras',
       'Azcapotzalco', 'Tláhuac', 'Milpa Alta'], dtype=object)

In [30]:
## Precio pronedio por delegación
df_copia.groupby("colonia")["precio"].mean().sort_values(ascending = False)

colonia
Tlalpan                   2493.466454
Cuajimalpa de Morelos     2150.578797
Cuauhtémoc                2126.383364
Álvaro Obregón            2117.761084
Miguel Hidalgo            1876.386781
Milpa Alta                1420.363636
Coyoacán                  1245.479152
La Magdalena Contreras    1195.074074
Benito Juárez             1120.261867
Xochimilco                1114.150000
Iztacalco                  930.181406
Venustiano Carranza        845.509804
Gustavo A. Madero          821.398693
Iztapalapa                 752.996169
Azcapotzalco               708.098837
Tláhuac                    403.411765
Name: precio, dtype: float64

In [31]:
## Por delegación contar cuantas noches mínimas permanece una persona
df_copia.groupby("colonia")["noches_minimas"].mean().sort_values(ascending = False)

colonia
Tláhuac                   12.150000
Cuajimalpa de Morelos      8.553659
Tlalpan                    6.820099
La Magdalena Contreras     6.152174
Álvaro Obregón             5.260736
Benito Juárez              5.247813
Azcapotzalco               5.082687
Miguel Hidalgo             4.947955
Coyoacán                   4.353793
Cuauhtémoc                 4.251478
Iztapalapa                 4.132450
Xochimilco                 3.206452
Gustavo A. Madero          2.920388
Venustiano Carranza        2.866593
Iztacalco                  2.122407
Milpa Alta                 1.500000
Name: noches_minimas, dtype: float64

In [32]:
## Precio promedio por tipo de habitación por cada colonia
#tipo_habitacion  
df_copia.groupby(['colonia','tipo_habitacion'])['precio'].mean()
## Manejo de Nan,  eliminar registros con nan, imputar valores del precio(moda, media, mediana, promedio por colonia y tipo de habitación,)

colonia                 tipo_habitacion
Azcapotzalco            Entire home/apt      863.431280
                        Private room         461.669173
Benito Juárez           Entire home/apt     1444.690254
                        Hotel room                  NaN
                        Private room         657.923695
                        Shared room          370.117647
Coyoacán                Entire home/apt     1706.249300
                        Hotel room          6806.000000
                        Private room         768.011461
                        Shared room          606.500000
Cuajimalpa de Morelos   Entire home/apt     2349.114286
                        Private room        1682.875000
Cuauhtémoc              Entire home/apt     2134.103544
                        Hotel room         58952.902439
                        Private room        1375.176598
                        Shared room          344.460227
Gustavo A. Madero       Entire home/apt     1009.534247
        

In [33]:
df_copia.columns

Index(['id', 'nombre', 'id_anfitrion', 'nombre_anfitrion', 'colonia',
       'latitud', 'longitud', 'tipo_habitacion', 'precio', 'noches_minimas',
       'numero_resenas', 'ultima_resena', 'resenas_por_mes',
       'propiedades_anfitrion', 'disponibilidad_365', 'number_of_reviews_ltm'],
      dtype='object')

In [34]:
# Entender si number_of_reviews influye en el precio del alojamiento
#number_of_reviews_ltm
df_copia.groupby("number_of_reviews_ltm")["precio"].mean().sort_values(ascending = False)


number_of_reviews_ltm
0      3220.883633
192    2530.000000
153    2427.500000
156    2220.000000
366    2210.000000
          ...     
109     632.333333
178     621.000000
117     589.166667
170     462.000000
147            NaN
Name: precio, Length: 191, dtype: float64

In [35]:
df_copia.groupby("colonia")["number_of_reviews_ltm"].mean().sort_values(ascending = False)

colonia
Venustiano Carranza       23.228225
Cuauhtémoc                18.583506
Azcapotzalco              17.891473
Iztacalco                 17.775934
Miguel Hidalgo            14.325388
Gustavo A. Madero         14.225243
Benito Juárez             12.550049
Coyoacán                  11.775333
Cuajimalpa de Morelos     11.151220
Iztapalapa                10.672185
Álvaro Obregón             8.369121
Tlalpan                    7.038462
Xochimilco                 5.780645
La Magdalena Contreras     4.586957
Milpa Alta                 3.800000
Tláhuac                    3.350000
Name: number_of_reviews_ltm, dtype: float64

In [36]:
 df_copia.groupby("colonia")["disponibilidad_365"].mean().sort_values(ascending = False)

colonia
Milpa Alta                282.700000
Xochimilco                275.741935
Tláhuac                   272.875000
Iztapalapa                264.135762
Gustavo A. Madero         257.376699
Cuajimalpa de Morelos     245.034146
Azcapotzalco              242.801034
Álvaro Obregón            237.968303
Iztacalco                 234.263485
Miguel Hidalgo            234.240761
Cuauhtémoc                232.449177
Venustiano Carranza       229.916207
Benito Juárez             228.702948
Tlalpan                   227.830025
Coyoacán                  216.537348
La Magdalena Contreras    200.615942
Name: disponibilidad_365, dtype: float64

In [37]:
df_copia["precio"]

0         3673.0
1        18000.0
2          591.0
3            NaN
4            NaN
          ...   
27046     1096.0
27047      260.0
27048      996.0
27049      458.0
27050     3082.0
Name: precio, Length: 27051, dtype: float64

In [38]:
#Manejo de Nan de la columna precio
filtro = df_copia['precio'].isna()
filtro

0        False
1        False
2        False
3         True
4         True
         ...  
27046    False
27047    False
27048    False
27049    False
27050    False
Name: precio, Length: 27051, dtype: bool

In [39]:
df_nan_precio = df_copia[filtro]
df_nan_precio

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,Cuauhtémoc,19.411520,-99.168570,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,Coyoacán,19.354480,-99.162170,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8
5,131610,MARIA DEL ALMA,647454,Fernando,Coyoacán,19.354160,-99.164880,Private room,NaN,1,0,NaT,NaN,3,0,0
9,187030,Spacious apt with the best rooftop in Condesa,899360,Julian,Cuauhtémoc,19.410570,-99.177300,Entire home/apt,NaN,4,103,2023-09-08,0.61,6,0,0
10,187745,Extraordinarily Spacious Apt. in Condesa,899360,Julian,Cuauhtémoc,19.410570,-99.177300,Entire home/apt,NaN,14,24,2018-01-03,0.17,6,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26375,1492790535447381248,3 High-End Apts with Private Rooftop @ Roma Norte,114564195,Fernando,Cuauhtémoc,19.420714,-99.165345,Entire home/apt,NaN,1,0,NaT,NaN,28,51,0
26399,1493527488230562415,Muy práctico y bonito,714991006,Claudia,Benito Juárez,19.384030,-99.137250,Entire home/apt,NaN,1,2,2025-09-05,2.00,1,267,2
26421,1494433329501631014,Habitación Plaza de toros México,385040544,Francisco,Benito Juárez,19.382454,-99.184501,Private room,NaN,1,0,NaT,NaN,1,364,0
26452,1495750044495995126,Espacio ejecutivo en el centro,23021265,Jesy,Cuauhtémoc,19.429344,-99.158556,Entire home/apt,NaN,1,1,2025-08-29,1.00,27,326,1


In [40]:
# Cuantos valores nan en precio se tiene por colonia
df_nan_precio.groupby("colonia").count()

,id,nombre,id_anfitrion,nombre_anfitrion,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
colonia,,,,,,,,,,,,,,,
Azcapotzalco,43,43,43,43,43,43,43,0,43,43,31,31,43,43,43
Benito Juárez,559,559,559,558,559,559,559,0,559,559,416,416,559,559,559
Coyoacán,312,312,312,311,312,312,312,0,312,312,220,220,312,312,312
Cuajimalpa de Morelos,61,61,61,61,61,61,61,0,61,61,38,38,61,61,61
Cuauhtémoc,1261,1261,1261,1261,1261,1261,1261,0,1261,1261,1043,1043,1261,1261,1261
Gustavo A. Madero,56,56,56,56,56,56,56,0,56,56,34,34,56,56,56
Iztacalco,41,41,41,41,41,41,41,0,41,41,32,32,41,41,41
Iztapalapa,41,41,41,41,41,41,41,0,41,41,16,16,41,41,41
La Magdalena Contreras,30,30,30,30,30,30,30,0,30,30,16,16,30,30,30


In [41]:
df_nan_precio["colonia"].value_counts(normalize =  True)

colonia
Cuauhtémoc                0.361940
Miguel Hidalgo            0.170494
Benito Juárez             0.160448
Coyoacán                  0.089552
Tlalpan                   0.051665
Álvaro Obregón            0.047646
Venustiano Carranza       0.026119
Cuajimalpa de Morelos     0.017509
Gustavo A. Madero         0.016073
Azcapotzalco              0.012342
Iztapalapa                0.011768
Iztacalco                 0.011768
Xochimilco                0.010046
La Magdalena Contreras    0.008611
Milpa Alta                0.002296
Tláhuac                   0.001722
Name: proportion, dtype: float64

In [42]:
## Imputación de valor nan en la columna precio

In [43]:
# Remplazo por promedio,promedio,  moda, mediana, (tomando en cuenta todos los precios del dataset)
# Remplazo por promedio,promedio, moda, mediana, pero por colonia
# Remplazo por promedio, media, moda  pero por colonia y tipo de propiedad  (Imputar primero los valores del tipo de propiedad)
# Remplazo por promedio,promedio, moda, mediana, de acuerdo al tipo de propiedad

#Promedio suaviza el precio, 


In [44]:
# El precio para el alojamiento de casas grandes NO es el mismo en cada colonia
df_copia.groupby("tipo_habitacion")["precio"].mean().sort_values(ascending = False)

tipo_habitacion
Hotel room         42841.000000
Entire home/apt     1930.279504
Private room        1226.560462
Shared room          360.885827
Name: precio, dtype: float64

In [45]:
## Valores atípicos en precio

In [46]:
df_copia["precio"].min()

61.0

In [47]:
df_copia["precio"].max()

900000.0

In [48]:
#Registro con el precio mas alto
filtro =  df_copia['precio'] == 900000
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
14503,892676742760732948,¡Recámara acogedora y muy cómoda!,270168697,Evan,Tlalpan,19.28152,-99.15623,Private room,900000.0,1,3,2024-09-29,0.13,1,364,1


In [49]:
#Registro con el precio mas bajo
filtro =  df_copia['precio'] == 61
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
22296,1327580316106887410,JC Friends Roma,244230249,Joel Genaro,Cuauhtémoc,19.40455,-99.16268,Shared room,61.0,1,4,2025-06-01,0.62,9,73,4


In [50]:
q = df_copia["precio"].quantile(0.99)

In [51]:
df_copia[df_copia["precio"] < q]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm
0,35797,Villa Dante,153786,Dici,Cuajimalpa de Morelos,19.382830,-99.271780,Entire home/apt,3673.0,1,0,NaT,NaN,1,363,0
2,56074,Great space in historical San Rafael,265650,Maris,Cuauhtémoc,19.439770,-99.156050,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1
6,165772,BEST 5 Bedroom HOUSE IN S. Miguel Chapultepec,790208,Francisco Carlos,Miguel Hidalgo,19.408260,-99.186590,Entire home/apt,3673.0,2,386,2025-08-10,2.24,5,201,28
7,171109,Cool room near WTC and Metrobus,816295,Carlos,Benito Juárez,19.396750,-99.175810,Private room,321.0,4,123,2025-09-07,0.88,2,301,15
8,180808,"Huge Luxurious Suite 70's style, perfectly loc...",36836,Roberto,Cuauhtémoc,19.423900,-99.168900,Entire home/apt,1190.0,3,62,2025-09-01,0.38,1,315,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
27046,1518356968266486778,Suite Frida,674644167,Heyliana,Venustiano Carranza,19.442240,-99.113440,Private room,1096.0,1,0,NaT,NaN,5,365,0
27047,1518388105556721163,Habitación a un costado de Ciudad Universitaria,222854336,Marco Antonio,Coyoacán,19.308017,-99.168158,Private room,260.0,1,0,NaT,NaN,1,365,0
27048,1518433354720281854,Espacioso departamento en Anzures,470800231,Hector Fernando,Miguel Hidalgo,19.434460,-99.174010,Entire home/apt,996.0,1,0,NaT,NaN,20,365,0
27049,1518462568893975567,Casa Roma Baja 05,710857718,Camila,Cuauhtémoc,19.406435,-99.160934,Entire home/apt,458.0,1,0,NaT,NaN,7,363,0


## Imputación de la columna precio
Se usa el promedio por que suaviza los precios, 
Desicion: Remplazo por promedio por colonia y tipo de propiedad para tener un precio más acorde a la ubicación, ya que como se analizo este depende de la colonia y de tipo de propiedad.


In [52]:
df_copia["precio_agregado"] = df_copia["precio"]
df_precios = df_copia.groupby("colonia")["precio"].mean()
df_precios




colonia
Azcapotzalco               708.098837
Benito Juárez             1120.261867
Coyoacán                  1245.479152
Cuajimalpa de Morelos     2150.578797
Cuauhtémoc                2126.383364
Gustavo A. Madero          821.398693
Iztacalco                  930.181406
Iztapalapa                 752.996169
La Magdalena Contreras    1195.074074
Miguel Hidalgo            1876.386781
Milpa Alta                1420.363636
Tlalpan                   2493.466454
Tláhuac                    403.411765
Venustiano Carranza        845.509804
Xochimilco                1114.150000
Álvaro Obregón            2117.761084
Name: precio, dtype: float64

In [53]:
## Diccionario con promedios por cada colonoa --> Python
## Groupby

df_copia["precio_agregado"] = np.where(df_copia["precio_agregado"].isna(), df_copia["colonia"].map(df_precios), df_copia["precio_agregado"])

In [54]:

df_copia["precio_agregado"] 

0         3673.000000
1        18000.000000
2          591.000000
3         2126.383364
4         1245.479152
             ...     
27046     1096.000000
27047      260.000000
27048      996.000000
27049      458.000000
27050     3082.000000
Name: precio_agregado, Length: 27051, dtype: float64

In [55]:

df_copia

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado
0,35797,Villa Dante,153786,Dici,Cuajimalpa de Morelos,19.382830,-99.271780,Entire home/apt,3673.0,1,0,NaT,NaN,1,363,0,3673.000000
1,44616,Condesa Haus,196253,Fernando,Cuauhtémoc,19.411620,-99.177940,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1,18000.000000
2,56074,Great space in historical San Rafael,265650,Maris,Cuauhtémoc,19.439770,-99.156050,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1,591.000000
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,Cuauhtémoc,19.411520,-99.168570,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1,2126.383364
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,Coyoacán,19.354480,-99.162170,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8,1245.479152
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
27046,1518356968266486778,Suite Frida,674644167,Heyliana,Venustiano Carranza,19.442240,-99.113440,Private room,1096.0,1,0,NaT,NaN,5,365,0,1096.000000
27047,1518388105556721163,Habitación a un costado de Ciudad Universitaria,222854336,Marco Antonio,Coyoacán,19.308017,-99.168158,Private room,260.0,1,0,NaT,NaN,1,365,0,260.000000
27048,1518433354720281854,Espacioso departamento en Anzures,470800231,Hector Fernando,Miguel Hidalgo,19.434460,-99.174010,Entire home/apt,996.0,1,0,NaT,NaN,20,365,0,996.000000
27049,1518462568893975567,Casa Roma Baja 05,710857718,Camila,Cuauhtémoc,19.406435,-99.160934,Entire home/apt,458.0,1,0,NaT,NaN,7,363,0,458.000000


In [56]:
df_copia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27051 entries, 0 to 27050
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     27051 non-null  int64         
 1   nombre                 27051 non-null  object        
 2   id_anfitrion           27051 non-null  int64         
 3   nombre_anfitrion       27045 non-null  object        
 4   colonia                27051 non-null  object        
 5   latitud                27051 non-null  float64       
 6   longitud               27051 non-null  float64       
 7   tipo_habitacion        27051 non-null  object        
 8   precio                 23567 non-null  float64       
 9   noches_minimas         27051 non-null  int64         
 10  numero_resenas         27051 non-null  int64         
 11  ultima_resena          23650 non-null  datetime64[ns]
 12  resenas_por_mes        23650 non-null  float64       
 13  p

In [57]:
df_copia.groupby(['colonia','tipo_habitacion'])['precio'].mean()

colonia                 tipo_habitacion
Azcapotzalco            Entire home/apt      863.431280
                        Private room         461.669173
Benito Juárez           Entire home/apt     1444.690254
                        Hotel room                  NaN
                        Private room         657.923695
                        Shared room          370.117647
Coyoacán                Entire home/apt     1706.249300
                        Hotel room          6806.000000
                        Private room         768.011461
                        Shared room          606.500000
Cuajimalpa de Morelos   Entire home/apt     2349.114286
                        Private room        1682.875000
Cuauhtémoc              Entire home/apt     2134.103544
                        Hotel room         58952.902439
                        Private room        1375.176598
                        Shared room          344.460227
Gustavo A. Madero       Entire home/apt     1009.534247
        

In [58]:
df_copia.groupby(['colonia','tipo_habitacion'])['precio_agregado'].mean()

colonia                 tipo_habitacion
Azcapotzalco            Entire home/apt      851.849475
                        Private room         501.965848
Benito Juárez           Entire home/apt     1389.958717
                        Hotel room          1120.261867
                        Private room         750.391329
                        Shared room          391.550339
Coyoacán                Entire home/apt     1654.670552
                        Hotel room          2635.609364
                        Private room         882.041368
                        Shared room          606.500000
Cuajimalpa de Morelos   Entire home/apt     2320.652904
                        Private room        1758.311096
Cuauhtémoc              Entire home/apt     2133.456702
                        Hotel room         45272.444143
                        Private room        1488.318476
                        Shared room          354.527590
Gustavo A. Madero       Entire home/apt      996.725017
        

In [59]:
#Imputación considerando colonia y tipo de habitación
df_precios_col_hab = df_copia.groupby(["colonia","tipo_habitacion"])["precio"].mean()
df_precios_col_hab

colonia                 tipo_habitacion
Azcapotzalco            Entire home/apt      863.431280
                        Private room         461.669173
Benito Juárez           Entire home/apt     1444.690254
                        Hotel room                  NaN
                        Private room         657.923695
                        Shared room          370.117647
Coyoacán                Entire home/apt     1706.249300
                        Hotel room          6806.000000
                        Private room         768.011461
                        Shared room          606.500000
Cuajimalpa de Morelos   Entire home/apt     2349.114286
                        Private room        1682.875000
Cuauhtémoc              Entire home/apt     2134.103544
                        Hotel room         58952.902439
                        Private room        1375.176598
                        Shared room          344.460227
Gustavo A. Madero       Entire home/apt     1009.534247
        

In [60]:
filtro = df_precios_col_hab.isna()
df_habitacion_nan = df_precios_col_hab[filtro]
df_habitacion_nan

colonia        tipo_habitacion
Benito Juárez  Hotel room        NaN
Tlalpan        Shared room       NaN
Name: precio, dtype: float64

In [61]:
#df_copia["precio_agregado2"] = df_copia["precio"]

In [62]:
promedio_general = df_copia.groupby("colonia")["precio"].mean()
promedio_general
filtro1 = ~df_copia["precio"].isna()
df1 = df_copia[filtro1]
df1

filtro1 = (df1["colonia"] == "Benito Juárez") & (df1["tipo_habitacion"] == "Hotel room" )
df1[filtro1]

#print("Promedio: ", promedio)

#df_valores = df_copia[filtro1]
#df_valores

#filtro2 = (df_valores["colonia"] == "Benito Juárez") & (df_valores["tipo_habitacion"] == "Hotel room")


#print("Promedio para Benito Juárez: ", promedio)

#print(promedio_benito_juarez)
#df_copia["precio_agregado2"] = np.where( (df_copia["colonia"]=="Benito Juárez" & df_copia["tipo_habitacion"] == "Hotel room") , )

#mean(skipna=True)

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado


## Uso de transform

In [63]:
alumnos = pd.DataFrame({
    "grupo": ["A", "A", "A", "B", "B", "B"],
    "nombre": ["Ana", "Luis", "Marta", "Pedro", "Sofía", "Juan"],
    "calificacion": [8, 10, 6, 7, 9, 5]
})

alumnos

,grupo,nombre,calificacion
0,A,Ana,8
1,A,Luis,10
2,A,Marta,6
3,B,Pedro,7
4,B,Sofía,9
5,B,Juan,5


In [64]:
promedio  = alumnos.groupby("grupo")["calificacion"].mean()
promedio  

promedio.index

Index(['A', 'B'], dtype='object', name='grupo')

In [65]:
alumnos["promedio_grupo"] = (
    alumnos.groupby("grupo")["calificacion"]
    .transform("mean")
)


In [66]:
alumnos["promedio_grupo"] = (
    alumnos.groupby("grupo")["calificacion"]
    .transform("mean")
)
alumnos

,grupo,nombre,calificacion,promedio_grupo
0,A,Ana,8,8.0
1,A,Luis,10,8.0
2,A,Marta,6,8.0
3,B,Pedro,7,7.0
4,B,Sofía,9,7.0
5,B,Juan,5,7.0


Transform coloca la mediana en cada fila correspondiente

## Imputación de la columna `precio`

La columna `precio` tiene valores faltantes. Como el precio de un alojamiento depende mucho de la ubicación(dada por la ubicaciòn) y del tipo de habitación, Por lo tanto no serìa adecuado que se imputen  los  valores faltantes con el promedio general, ya que meterìa sesgo en nuestros datos.

Despuès de hacer un anàlisis en las variables, podemos 

1. Primero, mediana por `colonia` y `tipo_habitacion`
2. Si todavía aùn hay faltantes, mediana por `colonia`
3. Si todavía hay faltantes, mediana por `tipo_habitacion`
4. Finalmente se puede usar mediana general del dataset

Se usa la **mediana** porque el precio tiene valores atípicos muy altos; la media puede deformarse por esos casos extremos.


In [67]:
# 1. Contar valores faltantes antes de la imputación
faltantes_antes = df_copia["precio"].isna().sum()
faltantes_antes


np.int64(3484)

In [68]:
porcentaje_faltantes_antes = df_copia["precio"].isna().mean() * 100
porcentaje_faltantes_antes

np.float64(12.879375993493772)

In [69]:
df_copia["precio_imputado"] = df_copia["precio"]

`transform()` calcula un resumen por grupo, pero devuelve un resultado del mismo tamaño que el DataFrame original.

Por ejemplo, si una fila pertenece a la colonia Roma y al tipo `Alojamiento entero`, `transform("median")` coloca en esa fila la mediana del precio de ese grupo. Por eso se puede usar directamente dentro de `fillna()`.

In [70]:
df_copia[[
    "colonia",
    "tipo_habitacion",
    "precio",
    "precio_imputado"
]].head(10)

,colonia,tipo_habitacion,precio,precio_imputado
0,Cuajimalpa de Morelos,Entire home/apt,3673.0,3673.0
1,Cuauhtémoc,Entire home/apt,18000.0,18000.0
2,Cuauhtémoc,Entire home/apt,591.0,591.0
3,Cuauhtémoc,Entire home/apt,NaN,NaN
4,Coyoacán,Entire home/apt,NaN,NaN
5,Coyoacán,Private room,NaN,NaN
6,Miguel Hidalgo,Entire home/apt,3673.0,3673.0
7,Benito Juárez,Private room,321.0,321.0
8,Cuauhtémoc,Entire home/apt,1190.0,1190.0
9,Cuauhtémoc,Entire home/apt,NaN,NaN


In [71]:
# 1. Calcular la mediana del precio para cada colonia y tipo de habitación
mediana_por_grupo = df_copia.groupby(
    ["colonia", "tipo_habitacion"]
)["precio"].transform("median")

mediana_por_grupo

0        1567.0
1        1317.0
2        1317.0
3        1317.0
4        1061.0
          ...  
27046     525.0
27047     520.0
27048    1500.0
27049    1317.0
27050    1317.0
Name: precio, Length: 27051, dtype: float64

In [72]:
#  Rellenar los precios faltantes usando esa mediana
df_copia["precio_imputado"] = df_copia["precio_imputado"].fillna(
    mediana_por_grupo
)
df_copia["precio_imputado"] 

0         3673.0
1        18000.0
2          591.0
3         1317.0
4         1061.0
          ...   
27046     1096.0
27047      260.0
27048      996.0
27049      458.0
27050     3082.0
Name: precio_imputado, Length: 27051, dtype: float64

In [73]:
faltantes_despues_nivel_1 = df_copia["precio_imputado"].isna().sum()

faltantes_despues_nivel_1

np.int64(6)

## Imputar los faltantes restantes usando la mediana por tipo de habitación

In [74]:

df_copia["precio_imputado"] = df_copia["precio_imputado"].fillna(
    df_copia.groupby("tipo_habitacion")["precio"].transform("median")
)

# Contar cuántos valores siguen faltando después del nivel 3
faltantes_despues_nivel_3 = df_copia["precio_imputado"].isna().sum()

faltantes_despues_nivel_3

np.int64(0)

In [75]:
df_copia

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado,precio_imputado
0,35797,Villa Dante,153786,Dici,Cuajimalpa de Morelos,19.382830,-99.271780,Entire home/apt,3673.0,1,0,NaT,NaN,1,363,0,3673.000000,3673.0
1,44616,Condesa Haus,196253,Fernando,Cuauhtémoc,19.411620,-99.177940,Entire home/apt,18000.0,1,65,2025-01-01,0.38,9,360,1,18000.000000,18000.0
2,56074,Great space in historical San Rafael,265650,Maris,Cuauhtémoc,19.439770,-99.156050,Entire home/apt,591.0,15,84,2025-02-27,0.48,1,333,1,591.000000,591.0
3,67703,"2 bedroom apt. deco bldg, Condesa",334451,Nicholas,Cuauhtémoc,19.411520,-99.168570,Entire home/apt,NaN,2,50,2024-10-30,0.30,2,252,1,2126.383364,1317.0
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,Coyoacán,19.354480,-99.162170,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8,1245.479152,1061.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
27046,1518356968266486778,Suite Frida,674644167,Heyliana,Venustiano Carranza,19.442240,-99.113440,Private room,1096.0,1,0,NaT,NaN,5,365,0,1096.000000,1096.0
27047,1518388105556721163,Habitación a un costado de Ciudad Universitaria,222854336,Marco Antonio,Coyoacán,19.308017,-99.168158,Private room,260.0,1,0,NaT,NaN,1,365,0,260.000000,260.0
27048,1518433354720281854,Espacioso departamento en Anzures,470800231,Hector Fernando,Miguel Hidalgo,19.434460,-99.174010,Entire home/apt,996.0,1,0,NaT,NaN,20,365,0,996.000000,996.0
27049,1518462568893975567,Casa Roma Baja 05,710857718,Camila,Cuauhtémoc,19.406435,-99.160934,Entire home/apt,458.0,1,0,NaT,NaN,7,363,0,458.000000,458.0


## Funciòn `agg()` : Signfifica agregar o resumir datos

- `agg()` permite obtener resúmenes por grupos.
- Calcular varias medidas al mismo tiempo

Se puede obtener información como:

- ¿Cuál es el precio promedio por colonia?
- ¿Cuál es el precio mínimo y máximo por tipo de habitación?
- ¿Cuántos alojamientos hay en cada colonia?
- ¿Cuál es la disponibilidad promedio por colonia?

## Ejemplo: resumen por colonia

```python
resumen_colonia = datos.groupby("colonia").agg(
    alojamientos=("nombre", "count"),
    precio_promedio=("precio", "mean"),
    precio_mediano=("precio", "median"),
    precio_minimo=("precio", "min"),
    precio_maximo=("precio", "max")
).reset_index()

resumen_colonia.head()

In [76]:
resumen_colonia = df_copia.groupby("colonia").agg(
    alojamientos=("nombre", "count"),
    precio_promedio=("precio", "mean"),
    precio_mediana=("precio", "median"),
    precio_minimo=("precio", "min"),
    precio_maximo=("precio", "max")
).reset_index()

resumen_colonia.head()

,colonia,alojamientos,precio_promedio,precio_mediana,precio_minimo,precio_maximo
0,Azcapotzalco,387,708.098837,511.5,191.0,8198.0
1,Benito Juárez,3087,1120.261867,811.0,163.0,50000.0
2,Coyoacán,1727,1245.479152,794.0,152.0,159200.0
3,Cuajimalpa de Morelos,410,2150.578797,1286.0,180.0,49500.0
4,Cuauhtémoc,12514,2126.383364,1185.0,61.0,734464.0


## Ejercicio: 
- ¿Cuántos alojamientos hay por tipo de habitación?
- ¿Cuál es el precio promedio de cada tipo?
- ¿Cuál es el precio mediano?
- ¿Cuál es la disponibilidad promedio durante el año?
- ¿Cuál es el número promedio de reseñas?

In [77]:
resumen_colonia = df_copia.groupby("tipo_habitacion").agg(
    alojamientos=("nombre", "count"),
    precio_promedio=("precio", "mean"),
    precio_mediana=("precio", "median"),
    precio_minimo=("precio", "min"),
    precio_maximo=("precio", "max")
).reset_index()

resumen_colonia.head()

,tipo_habitacion,alojamientos,precio_promedio,precio_mediana,precio_minimo,precio_maximo
0,Entire home/apt,17713,1930.279504,1250.0,150.0,347429.0
1,Hotel room,86,42841.000000,2000.0,415.0,734464.0
2,Private room,8995,1226.560462,580.0,137.0,900000.0
3,Shared room,257,360.885827,261.0,61.0,12841.0


## Multiíndices en Pandas



Un **multiíndice** es un índice con varios niveles.

En Pandas aparece comúnmente cuando agrupamos los datos por más de una columna usando `groupby()`.

## Idea principal

Cuando agrupamos por una sola columna, el índice tiene un solo nivel.


In [78]:

resumen_colonia = df_copia.groupby("colonia").agg(
    total_alojamientos=("nombre", "count"),
    precio_promedio=("precio", "mean")
)

resumen_colonia.head()

,total_alojamientos,precio_promedio
colonia,,
Azcapotzalco,387,708.098837
Benito Juárez,3087,1120.261867
Coyoacán,1727,1245.479152
Cuajimalpa de Morelos,410,2150.578797
Cuauhtémoc,12514,2126.383364


In [79]:
resumen_colonia.index

Index(['Azcapotzalco', 'Benito Juárez', 'Coyoacán', 'Cuajimalpa de Morelos',
       'Cuauhtémoc', 'Gustavo A. Madero', 'Iztacalco', 'Iztapalapa',
       'La Magdalena Contreras', 'Miguel Hidalgo', 'Milpa Alta', 'Tlalpan',
       'Tláhuac', 'Venustiano Carranza', 'Xochimilco', 'Álvaro Obregón'],
      dtype='object', name='colonia')

In [80]:
print(resumen_colonia.index)

Index(['Azcapotzalco', 'Benito Juárez', 'Coyoacán', 'Cuajimalpa de Morelos',
       'Cuauhtémoc', 'Gustavo A. Madero', 'Iztacalco', 'Iztapalapa',
       'La Magdalena Contreras', 'Miguel Hidalgo', 'Milpa Alta', 'Tlalpan',
       'Tláhuac', 'Venustiano Carranza', 'Xochimilco', 'Álvaro Obregón'],
      dtype='object', name='colonia')


In [81]:
resumen_colonia.head()

,total_alojamientos,precio_promedio
colonia,,
Azcapotzalco,387,708.098837
Benito Juárez,3087,1120.261867
Coyoacán,1727,1245.479152
Cuajimalpa de Morelos,410,2150.578797
Cuauhtémoc,12514,2126.383364


In [82]:
resumen = df_copia.groupby("colonia").agg(
    precio_promedio=("precio", "mean")
)

resumen.head()

,precio_promedio
colonia,
Azcapotzalco,708.098837
Benito Juárez,1120.261867
Coyoacán,1245.479152
Cuajimalpa de Morelos,2150.578797
Cuauhtémoc,2126.383364


In [83]:
resumen.index

Index(['Azcapotzalco', 'Benito Juárez', 'Coyoacán', 'Cuajimalpa de Morelos',
       'Cuauhtémoc', 'Gustavo A. Madero', 'Iztacalco', 'Iztapalapa',
       'La Magdalena Contreras', 'Miguel Hidalgo', 'Milpa Alta', 'Tlalpan',
       'Tláhuac', 'Venustiano Carranza', 'Xochimilco', 'Álvaro Obregón'],
      dtype='object', name='colonia')

## Multiíndice

Sirve para organizar datos en varios niveles

- Nivel 1 del índice: colonia
- Nivel 2 del índice: tipo_habitacion


In [84]:
resumen = df_copia.groupby(["colonia", "tipo_habitacion"]).agg(
    precio_promedio=("precio", "mean")
)

print(resumen)

resumen.index

                                        precio_promedio
colonia                tipo_habitacion                 
Azcapotzalco           Entire home/apt       863.431280
                       Private room          461.669173
Benito Juárez          Entire home/apt      1444.690254
                       Hotel room                   NaN
                       Private room          657.923695
                       Shared room           370.117647
Coyoacán               Entire home/apt      1706.249300
                       Hotel room           6806.000000
                       Private room          768.011461
                       Shared room           606.500000
Cuajimalpa de Morelos  Entire home/apt      2349.114286
                       Private room         1682.875000
Cuauhtémoc             Entire home/apt      2134.103544
                       Hotel room          58952.902439
                       Private room         1375.176598
                       Shared room           344

MultiIndex([(          'Azcapotzalco', 'Entire home/apt'),
            (          'Azcapotzalco',    'Private room'),
            (         'Benito Juárez', 'Entire home/apt'),
            (         'Benito Juárez',      'Hotel room'),
            (         'Benito Juárez',    'Private room'),
            (         'Benito Juárez',     'Shared room'),
            (              'Coyoacán', 'Entire home/apt'),
            (              'Coyoacán',      'Hotel room'),
            (              'Coyoacán',    'Private room'),
            (              'Coyoacán',     'Shared room'),
            ( 'Cuajimalpa de Morelos', 'Entire home/apt'),
            ( 'Cuajimalpa de Morelos',    'Private room'),
            (            'Cuauhtémoc', 'Entire home/apt'),
            (            'Cuauhtémoc',      'Hotel room'),
            (            'Cuauhtémoc',    'Private room'),
            (            'Cuauhtémoc',     'Shared room'),
            (     'Gustavo A. Madero', 'Entire home/apt'

In [85]:
resumen.sort_values("precio_promedio", ascending=False).head(10)

,,precio_promedio
colonia,tipo_habitacion,
Cuauhtémoc,Hotel room,58952.902439
Coyoacán,Hotel room,6806.000000
Tlalpan,Private room,3310.602941
Álvaro Obregón,Private room,2992.357143
Cuajimalpa de Morelos,Entire home/apt,2349.114286
Miguel Hidalgo,Entire home/apt,2180.229831
Cuauhtémoc,Entire home/apt,2134.103544
Milpa Alta,Entire home/apt,1873.000000
Coyoacán,Entire home/apt,1706.249300


Preguntas

1. ¿Qué colonia tiene el precio promedio más alto?
2. ¿Qué tipo de habitación suele ser más caro?
3. ¿Qué combinación de colonia y tipo de habitación tiene más alojamientos?
4. ¿Por qué fue útil usar groupby con dos columnas?

- En caso de querer converitr el resultado a una tabla normal usamos

- resumen = resumen.reset_index()

- Reseteamos el índice cuando queremos dejar de trabajar con una estructura jerárquica y volver a tener una tabla tradicional, donde cada variable es una columna.


In [86]:
resumen = resumen.reset_index()
resumen.sort_values("precio_promedio", ascending=False).head()

,colonia,tipo_habitacion,precio_promedio
13,Cuauhtémoc,Hotel room,58952.902439
7,Coyoacán,Hotel room,6806.000000
34,Tlalpan,Private room,3310.602941
45,Álvaro Obregón,Private room,2992.357143
10,Cuajimalpa de Morelos,Entire home/apt,2349.114286


## Uso de Pivot

## ¿Qué es una tabla dinámica o `pivot_table`?

Una tabla dinámica permite resumir y reorganizar un DataFrame para analizar la información de forma cruzada.

En lugar de ver los datos registro por registro, una tabla dinámica nos permite responder preguntas como:

**¿Cuál es el precio promedio de los alojamientos por colonia y tipo de habitación?**

En el dataset de Airbnb, cada fila representa un alojamiento. Algunas columnas importantes son:

- `colonia`: zona donde se encuentra el alojamiento.
- `tipo_habitacion`: tipo de alojamiento, por ejemplo habitación privada o departamento completo.
- `precio`: precio del alojamiento.

Con `pivot_table`, podemos convertir esta información en una tabla resumen donde:

- Las filas representan las colonias.
- Las columnas representan los tipos de habitación.
- Los valores representan el precio promedio.

La estructura general es:

```python
pd.pivot_table(
    data=df_copia,
    index="colonia",
    columns="tipo_habitacion",
    values="precio",
    aggfunc="mean"
)

In [87]:
pd.pivot_table(
    data=df_copia,
    index="colonia",
    columns="tipo_habitacion",
    values="precio_imputado",
    aggfunc="mean"
)

tipo_habitacion,Entire home/apt,Hotel room,Private room,Shared room
colonia,,,,
Azcapotzalco,856.241228,NaN,458.452830,NaN
Benito Juárez,1378.105438,2000.000000,634.338956,368.457143
Coyoacán,1634.019900,6806.000000,708.780807,606.500000
Cuajimalpa de Morelos,2236.993007,NaN,1532.895161,NaN
Cuauhtémoc,2065.641804,45784.648148,1266.105828,343.858757
Gustavo A. Madero,992.885106,NaN,640.075758,184.625000
Iztacalco,1092.833948,NaN,675.634615,1288.333333
Iztapalapa,737.827160,NaN,705.985612,402.000000
La Magdalena Contreras,1368.360656,NaN,872.389610,NaN


Esta tabla dinámica muestra el precio promedio de los alojamientos de Airbnb según dos variables:

- Filas: `colonia`
- Columnas: `tipo_habitacion`
- Valores: promedio de `precio`

Cada celda representa el precio promedio para una combinación específica.

Por ejemplo:

En `Benito Juárez`, el precio promedio de `Entire home/apt` es aproximadamente 1444.69, mientras que el de `Private room` es 657.92.

Los valores `NaN` significan que no existen registros para esa combinación. Por ejemplo, si aparece `NaN` en `Hotel room` para una colonia, significa que en esa colonia no hay alojamientos registrados de ese tipo.

Esta tabla permite comparar rápidamente qué colonias son más caras y cómo cambia el precio según el tipo de habitación.

# Ejercicios

- Ejercicio 1: Crear una tabla dinámica que muestre el precio promedio por colonia

In [88]:
pd.pivot_table(
    data=df_copia,
    index="colonia",
    values="precio_imputado",
    aggfunc="mean"
)

,precio_imputado
colonia,
Azcapotzalco,692.808786
Benito Juárez,1067.701328
Coyoacán,1153.526925
Cuajimalpa de Morelos,2024.046341
Cuauhtémoc,2021.657743
Gustavo A. Madero,786.916505
Iztacalco,914.014523
Iztapalapa,722.059603
La Magdalena Contreras,1091.623188


- Ejercicio 2: Crear una tabla dinámica que muestre el precio promedio por tipo de habitación.

In [89]:
pd.pivot_table(
    data=df_copia,
    index="tipo_habitacion",
    values="precio_imputado",
    aggfunc="mean"
)

,precio_imputado
tipo_habitacion,
Entire home/apt,1859.963925
Hotel room,29503.825581
Private room,1109.508171
Shared room,359.828794


- Ejercicio 4: Crear una tabla dinámica que muestre la mediana del precio por colonia y tipo de habitación.

In [90]:
pd.pivot_table(
    data=df_copia,
    index=["colonia","tipo_habitacion"],
    values="precio_imputado",
    aggfunc="median"
)

precio_imputado
colonia                tipo_habitacion                 
Azcapotzalco           Entire home/apt            767.0
                       Private room               442.0
Benito Juárez          Entire home/apt           1050.0
                       Hotel room                2000.0
                       Private room               540.0
                       Shared room                312.0
Coyoacán               Entire home/apt           1061.0
                       Hotel room                6806.0
                       Private room               520.0
                       Shared room                606.5
Cuajimalpa de Morelos  Entire home/apt           1567.0
                       Private room               753.0
Cuauhtémoc             Entire home/apt           1317.0
                       Hotel room                4254.0
                       Private room               651.0
                       Shared room                238.0
Gustavo A. Madero      Entire home/apt            765.0
                       Private room               400.0
                       Shared room                172.5
Iztacalco              Entire home/apt            945.0
                       Private room               525.0
                       Shared room               1378.0
Iztapalapa             Entire home/apt            590.0
                       Private room               495.0
                       Shared room                402.0
La Magdalena Contreras Entire home/apt           1076.0
                       Private room               630.0
Miguel Hidalgo         Entire home/apt           1500.0
                       Hotel room                 618.0
                       Private room               698.0
                       Shared room                266.0
Milpa Alta             Entire home/apt           1530.0
                       Private room               405.0
Tlalpan                Entire home/apt            900.0
                       Private room               497.0
                       Shared room                261.0
Tláhuac                Entire home/apt            385.0
                       Private room               308.0
Venustiano Carranza    Entire home/apt            846.0
                       Private room               525.0
                       Shared room                450.0
Xochimilco             Entire home/apt            720.0
                       Private room               508.0
Álvaro Obregón         Entire home/apt           1149.0
                       Hotel room                1800.0
                       Private room               500.0
                       Shared room                267.0

In [91]:
pd.pivot_table(
    data=df_copia,
    index="colonia",
    columns = "tipo_habitacion",
    values="precio_imputado",
    aggfunc="median"
)

tipo_habitacion,Entire home/apt,Hotel room,Private room,Shared room
colonia,,,,
Azcapotzalco,767.0,NaN,442.0,NaN
Benito Juárez,1050.0,2000.0,540.0,312.0
Coyoacán,1061.0,6806.0,520.0,606.5
Cuajimalpa de Morelos,1567.0,NaN,753.0,NaN
Cuauhtémoc,1317.0,4254.0,651.0,238.0
Gustavo A. Madero,765.0,NaN,400.0,172.5
Iztacalco,945.0,NaN,525.0,1378.0
Iztapalapa,590.0,NaN,495.0,402.0
La Magdalena Contreras,1076.0,NaN,630.0,NaN


- Ejercicio 5: Crear una tabla dinámica que muestre cuántos alojamientos existen por colonia y tipo de habitación.

In [92]:
pd.pivot_table(
    data=df_copia,
    index="colonia",
    columns="tipo_habitacion",
    values="precio_imputado",
    aggfunc="count"
)

tipo_habitacion,Entire home/apt,Hotel room,Private room,Shared room
colonia,,,,
Azcapotzalco,228.0,NaN,159.0,NaN
Benito Juárez,1802.0,5.0,1245.0,35.0
Coyoacán,804.0,4.0,917.0,2.0
Cuajimalpa de Morelos,286.0,NaN,124.0,NaN
Cuauhtémoc,9023.0,54.0,3260.0,177.0
Gustavo A. Madero,235.0,NaN,264.0,16.0
Iztacalco,271.0,NaN,208.0,3.0
Iztapalapa,162.0,NaN,139.0,1.0
La Magdalena Contreras,61.0,NaN,77.0,NaN


In [93]:
filtro = df_copia["tipo_habitacion"] == "Entire home/apt"
data_filtrado = df_copia[filtro]

pd.pivot_table(
    data=data_filtrado,
    index="colonia",
    columns="tipo_habitacion",
    values="precio_imputado",
    aggfunc={'precio_imputado': ["sum","mean","max"]}
)


 

,max,mean,sum
tipo_habitacion,Entire home/apt,Entire home/apt,Entire home/apt
colonia,,,
Azcapotzalco,8198.0,856.241228,195223.0
Benito Juárez,50000.0,1378.105438,2483346.0
Coyoacán,159200.0,1634.019900,1313752.0
Cuajimalpa de Morelos,49500.0,2236.993007,639780.0
Cuauhtémoc,347429.0,2065.641804,18638286.0
Gustavo A. Madero,9403.0,992.885106,233328.0
Iztacalco,5428.0,1092.833948,296158.0
Iztapalapa,4530.0,737.827160,119528.0


In [94]:
pd.pivot_table(
    data=df_copia,
    index="colonia",
    columns="tipo_habitacion",
    values="precio_imputado",
    aggfunc={'precio_imputado': ["sum","mean","max"]}
)



max                                      \
tipo_habitacion        Entire home/apt Hotel room Private room Shared room   
colonia                                                                      
Azcapotzalco                    8198.0        NaN       1120.0         NaN   
Benito Juárez                  50000.0     2000.0       7315.0      1448.0   
Coyoacán                      159200.0     6806.0       8263.0       813.0   
Cuajimalpa de Morelos          49500.0        NaN      44000.0         NaN   
Cuauhtémoc                    347429.0   734464.0     734464.0     12841.0   
Gustavo A. Madero               9403.0        NaN      20324.0       234.0   
Iztacalco                       5428.0        NaN       5680.0      1947.0   
Iztapalapa                      4530.0        NaN      24789.0       402.0   
La Magdalena Contreras          4490.0        NaN       5555.0         NaN   
Miguel Hidalgo                 65000.0     5619.0      11476.0      1200.0   
Milpa Alta                      6750.0        NaN        600.0         NaN   
Tlalpan                        67500.0        NaN     900000.0       261.0   
Tláhuac                          922.0        NaN        550.0         NaN   
Venustiano Carranza            10271.0        NaN       1688.0      1385.0   
Xochimilco                      8208.0        NaN       2029.0         NaN   
Álvaro Obregón                 21556.0     2228.0     765000.0       267.0   

                                  mean                             \
tipo_habitacion        Entire home/apt    Hotel room Private room   
colonia                                                             
Azcapotzalco                856.241228           NaN   458.452830   
Benito Juárez              1378.105438   2000.000000   634.338956   
Coyoacán                   1634.019900   6806.000000   708.780807   
Cuajimalpa de Morelos      2236.993007           NaN  1532.895161   
Cuauhtémoc                 2065.641804  45784.648148  1266.105828   
Gustavo A. Madero           992.885106           NaN   640.075758   
Iztacalco                  1092.833948           NaN   675.634615   
Iztapalapa                  737.827160           NaN   705.985612   
La Magdalena Contreras     1368.360656           NaN   872.389610   
Miguel Hidalgo             2104.967742    977.562500   952.336851   
Milpa Alta                 1815.833333           NaN   431.500000   
Tlalpan                    1427.908012           NaN  2541.070513   
Tláhuac                     471.857143           NaN   305.736842   
Venustiano Carranza        1020.998124           NaN   570.783934   
Xochimilco                 1377.142857           NaN   543.014085   
Álvaro Obregón             1462.939736   1727.571429  2407.589977   

                                                sum                          \
tipo_habitacion         Shared room Entire home/apt Hotel room Private room   
colonia                                                                       
Azcapotzalco                    NaN        195223.0        NaN      72894.0   
Benito Juárez            368.457143       2483346.0    10000.0     789752.0   
Coyoacán                 606.500000       1313752.0    27224.0     649952.0   
Cuajimalpa de Morelos           NaN        639780.0        NaN     190079.0   
Cuauhtémoc               343.858757      18638286.0  2472371.0    4127505.0   
Gustavo A. Madero        184.625000        233328.0        NaN     168980.0   
Iztacalco               1288.333333        296158.0        NaN     140532.0   
Iztapalapa               402.000000        119528.0        NaN      98132.0   
La Magdalena Contreras          NaN         83470.0        NaN      67174.0   
Miguel Hidalgo           458.875000       6982178.0    15641.0    1173279.0   
Milpa Alta                      NaN         32685.0        NaN       5178.0   
Tlalpan                  261.000000        481205.0        NaN    1189221.0   
Tláhuac                         NaN          9909.0    

-Ejercicio6: Comparar pivot_table con groupby.

Crear ambas soluciones para calcular:

precio promedio por colonia y tipo de habitación.

In [95]:
df_copia.groupby(["colonia","tipo_habitacion"])["precio"].mean()

colonia                 tipo_habitacion
Azcapotzalco            Entire home/apt      863.431280
                        Private room         461.669173
Benito Juárez           Entire home/apt     1444.690254
                        Hotel room                  NaN
                        Private room         657.923695
                        Shared room          370.117647
Coyoacán                Entire home/apt     1706.249300
                        Hotel room          6806.000000
                        Private room         768.011461
                        Shared room          606.500000
Cuajimalpa de Morelos   Entire home/apt     2349.114286
                        Private room        1682.875000
Cuauhtémoc              Entire home/apt     2134.103544
                        Hotel room         58952.902439
                        Private room        1375.176598
                        Shared room          344.460227
Gustavo A. Madero       Entire home/apt     1009.534247
        

In [96]:
pd.pivot_table(
    data=df_copia,
    index="colonia",
    columns="tipo_habitacion",
    values="precio_imputado",
    aggfunc="mean"
)

tipo_habitacion,Entire home/apt,Hotel room,Private room,Shared room
colonia,,,,
Azcapotzalco,856.241228,NaN,458.452830,NaN
Benito Juárez,1378.105438,2000.000000,634.338956,368.457143
Coyoacán,1634.019900,6806.000000,708.780807,606.500000
Cuajimalpa de Morelos,2236.993007,NaN,1532.895161,NaN
Cuauhtémoc,2065.641804,45784.648148,1266.105828,343.858757
Gustavo A. Madero,992.885106,NaN,640.075758,184.625000
Iztacalco,1092.833948,NaN,675.634615,1288.333333
Iztapalapa,737.827160,NaN,705.985612,402.000000
La Magdalena Contreras,1368.360656,NaN,872.389610,NaN


## Ejercicios regex con string

- Ejercicio 1: Buscar alojamientos que mencionan "loft"

In [97]:
df_copia.columns

Index(['id', 'nombre', 'id_anfitrion', 'nombre_anfitrion', 'colonia',
       'latitud', 'longitud', 'tipo_habitacion', 'precio', 'noches_minimas',
       'numero_resenas', 'ultima_resena', 'resenas_por_mes',
       'propiedades_anfitrion', 'disponibilidad_365', 'number_of_reviews_ltm',
       'precio_agregado', 'precio_imputado'],
      dtype='object')

In [98]:
filtro = df_copia["nombre"].str.contains(r"loft",
                                        case = False,
                                        na = False)
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado,precio_imputado
17,261897,Charming 1 bedroom loft w/ balcony in Roma/con...,1373924,Eric,Cuauhtémoc,19.418770,-99.171990,Entire home/apt,1018.0,5,354,2023-03-25,2.12,1,295,0,1018.000000,1018.0
62,822005,AMAZING LOFT in the heart of CDMX,4316391,Alejandro,Miguel Hidalgo,19.402430,-99.178280,Entire home/apt,NaN,4,55,2023-11-30,0.37,1,365,0,1876.386781,1500.0
71,939721,Loft with garden and private terrace in Roma N...,4821807,Alex,Cuauhtémoc,19.422880,-99.154880,Entire home/apt,1285.0,3,550,2025-09-02,3.62,3,286,16,1285.000000,1285.0
83,1288566,Luxury Loft-Style Apartment in Vibrant Polanco...,540705,Andrea,Miguel Hidalgo,19.431790,-99.194260,Entire home/apt,3185.0,1,228,2025-09-10,1.96,2,290,56,3185.000000,3185.0
89,1404363,SPOTLESS beautiful sunny loft in Del Valle,5341352,Maria Giselle,Benito Juárez,19.380150,-99.174020,Entire home/apt,760.0,4,151,2025-08-31,1.02,1,100,18,760.000000,760.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26940,1513250752581957039,El mejor Loft en la Roma. Céntrico. Equipado. ...,687634099,Maria Guadalupe,Cuauhtémoc,19.415750,-99.163840,Entire home/apt,1401.0,1,0,NaT,NaN,12,361,0,1401.000000,1401.0
26993,1516056089114079132,VH | Beautiful Condesa loft with patio | 50,665103179,Raul,Cuauhtémoc,19.407055,-99.172648,Entire home/apt,1303.0,1,1,2025-09-25,1.00,6,328,1,1303.000000,1303.0
26998,1516132650525205656,Renovado Loft en La del Valle,79648096,Celia,Benito Juárez,19.370264,-99.173148,Entire home/apt,734.0,2,0,NaT,NaN,3,310,0,734.000000,734.0
27006,1516298011226607114,Modern Loft with Balcony & Skyline Views in Roma,454330571,Lorenza,Cuauhtémoc,19.419209,-99.166040,Entire home/apt,1375.0,2,0,NaT,NaN,31,302,0,1375.000000,1375.0


- Ejercicio 2: Buscar alojamientos que mencionan "studio" o "estudio"

In [99]:
filtro = df_copia["nombre"].str.contains(r"studio|estudio",
                                        case = False,
                                        na = False)
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado,precio_imputado
4,70644,Beautiful light Studio Coyoacan- full equipped !,212109,Trisha,Coyoacán,19.354480,-99.162170,Entire home/apt,NaN,3,134,2025-08-18,0.81,3,234,8,1245.479152,1061.0
15,247543,"Independent studio, top location",1297488,Petra,Cuauhtémoc,19.435160,-99.138060,Entire home/apt,758.0,3,162,2025-09-16,0.97,2,304,28,758.000000,758.0
20,287940,Furnished lovely cozy studio,1497087,Penelope,Cuauhtémoc,19.412360,-99.175330,Entire home/apt,559.0,1,447,2025-09-06,3.05,13,328,21,559.000000,559.0
79,1167552,"Big private room with studio, Roma Mexico City.",6393641,José,Cuauhtémoc,19.413170,-99.157350,Private room,640.0,2,174,2025-09-10,1.16,1,241,24,640.000000,640.0
84,1309198,Linda y práctica habitación estudio,7116734,Mafer,Coyoacán,19.343560,-99.128600,Private room,450.0,2,2,2024-07-31,0.02,4,263,0,450.000000,450.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26804,1507430855289059624,Capitalia | Studio with Rooftop Jacuzzi & Parking,541272314,Capitalia,Cuauhtémoc,19.428454,-99.159846,Entire home/apt,1547.0,1,2,2025-09-24,2.00,22,342,2,1547.000000,1547.0
26912,1512028684852066581,Estudio Loft Equipado Six Flags CDMX,456019534,Lalo,Tlalpan,19.290945,-99.221719,Entire home/apt,793.0,2,0,NaT,NaN,7,346,0,793.000000,793.0
26918,1512540062930918635,Iluminado estudio con terraza,31540218,Roy,Cuauhtémoc,19.406567,-99.176086,Entire home/apt,672.0,2,0,NaT,NaN,12,78,0,672.000000,672.0
26958,1513928160113759832,1A Acogedor estudio con terraza privada Roma N...,333284391,Ana,Cuauhtémoc,19.418842,-99.155447,Entire home/apt,1654.0,2,0,NaT,NaN,2,347,0,1654.000000,1654.0


- Ejercicio 3: Buscar alojamientos que mencionan "metro"

In [100]:
filtro = df_copia["nombre"].str.contains(r"metro",
                                        case = False,
                                        na = False)
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado,precio_imputado
7,171109,Cool room near WTC and Metrobus,816295,Carlos,Benito Juárez,19.396750,-99.175810,Private room,321.0,4,123,2025-09-07,0.88,2,301,15,321.000000,321.0
49,642725,Recamara en depa a 5 minutos del metropolitano,3213044,Laura,Cuauhtémoc,19.408090,-99.181090,Private room,NaN,2,0,NaT,NaN,1,0,0,2126.383364,651.0
317,5593671,Chic Urban Spot by Food & Metro,29003246,Alfredo,Cuauhtémoc,19.407770,-99.177800,Entire home/apt,914.0,2,398,2025-09-15,3.10,1,304,34,914.000000,914.0
351,6204678,House near Condesa & metro for Foro Sol+Autodromo,29969396,Ricardo,Miguel Hidalgo,19.406020,-99.180980,Entire home/apt,NaN,2,138,2024-10-12,1.11,1,0,1,1876.386781,1500.0
492,7894945,XURU Confort I Apto movilidad reducida cerca M...,17468617,Rafael,Coyoacán,19.345840,-99.133510,Entire home/apt,2558.0,2,9,2024-02-11,0.25,14,365,0,2558.000000,2558.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25862,1476206887325885090,Conciertos o viajes: departamento a 350 del metro,497312274,Gabriela,Iztacalco,19.401840,-99.071680,Entire home/apt,1177.0,1,7,2025-09-25,4.12,1,337,7,1177.000000,1177.0
26330,1491159885372993696,Shared Room Near Metrobus,207240505,Luis,Cuauhtémoc,19.461120,-99.146700,Shared room,153.0,3,0,NaT,NaN,91,364,0,153.000000,153.0
26343,1491481504963912305,Shared room Metrobus San Simón,207240505,Luis,Cuauhtémoc,19.461450,-99.147200,Shared room,153.0,3,0,NaT,NaN,91,365,0,153.000000,153.0
26664,1501885077391100982,"Aeropuerto, bosques de Aragón, metro céntrico 777",593599865,Juan Carlos,Gustavo A. Madero,19.452860,-99.066440,Private room,322.0,1,2,2025-09-15,2.00,7,60,2,322.000000,322.0


- Ejercicio 4: Buscar nombres que terminan con "CDMX"

In [101]:
filtro = df_copia["nombre"].str.contains(r"CDMX$",
                                        case = False,
                                        na = False)
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado,precio_imputado
42,567176,Un descanso en CDMX,2336127,Ana Karen,Cuauhtémoc,19.455950,-99.159440,Private room,212.0,7,28,2023-11-03,0.23,1,302,0,212.000000,212.0
62,822005,AMAZING LOFT in the heart of CDMX,4316391,Alejandro,Miguel Hidalgo,19.402430,-99.178280,Entire home/apt,NaN,4,55,2023-11-30,0.37,1,365,0,1876.386781,1500.0
97,1565220,Descansa en la Suite de la Abuelita San Angel ...,2518152,Martha Laura,Álvaro Obregón,19.347350,-99.190160,Hotel room,1800.0,1,28,2024-01-02,0.19,7,323,0,1800.000000,1800.0
102,1727725,"Descansa en la suite del geógrafo, San Angel CDMX",2518152,Martha Laura,Álvaro Obregón,19.349230,-99.193070,Private room,2600.0,1,38,2025-05-09,0.26,7,355,2,2600.000000,2600.0
103,1736063,"Cómodo Estudio del Botánico, San Angel CDMX",2518152,Martha Laura,Álvaro Obregón,19.347230,-99.193960,Hotel room,2228.0,1,66,2025-03-23,0.47,7,0,8,2228.000000,2228.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
26893,1511194143128281269,"Stylish Flat in Trendy Roma Norte, CDMX",358250534,Inperio,Cuauhtémoc,19.421989,-99.169200,Entire home/apt,1026.0,1,1,2025-09-24,1.00,23,336,1,1026.000000,1026.0
26912,1512028684852066581,Estudio Loft Equipado Six Flags CDMX,456019534,Lalo,Tlalpan,19.290945,-99.221719,Entire home/apt,793.0,2,0,NaT,NaN,7,346,0,793.000000,793.0
26920,1512576774571975929,Acogedor depa con terraza en el corazón de la ...,9687996,Fabianna,Cuauhtémoc,19.432180,-99.167540,Entire home/apt,1402.0,5,0,NaT,NaN,1,15,0,1402.000000,1402.0
26974,1514727068185015787,"Cuartos con baño propio, céntricos en la CDMX",162327162,Julio,Azcapotzalco,19.473970,-99.180065,Private room,412.0,2,0,NaT,NaN,1,359,0,412.000000,412.0


- Ejercicio 5: Buscar nombres que empiezan con "Departament

In [102]:
filtro = df_copia["nombre"].str.contains(r"^Department",
                                        case = False,
                                        na = False)
df_copia[filtro]

,id,nombre,id_anfitrion,nombre_anfitrion,colonia,latitud,longitud,tipo_habitacion,precio,noches_minimas,numero_resenas,ultima_resena,resenas_por_mes,propiedades_anfitrion,disponibilidad_365,number_of_reviews_ltm,precio_agregado,precio_imputado
1908,19144158,"Department, studio type excellent location.",19975112,Jose Antonio,Benito Juárez,19.389200,-99.137500,Entire home/apt,583.0,2,114,2025-09-13,1.14,2,156,36,583.000000,583.0
1925,19267780,Department Bamboo/cuajimalpa,91356852,Sandra,Cuauhtémoc,19.446260,-99.129590,Private room,NaN,30,0,NaT,NaN,1,0,0,2126.383364,651.0
1929,19301189,Department Narvarte,134275953,Alejandro,Benito Juárez,19.399640,-99.157560,Entire home/apt,NaN,1,0,NaT,NaN,1,0,0,1120.261867,1050.0
6355,42410985,"Department in CDMX, north zone",15206663,Elias,Gustavo A. Madero,19.486440,-99.148560,Entire home/apt,1296.0,2,20,2025-09-21,0.37,3,0,12,1296.000000,1296.0
19286,1163813409613712505,Department Airport/Estadio GNP,442210468,Viridiana,Iztacalco,19.407699,-99.062594,Entire home/apt,NaN,1,100,2025-08-02,6.34,1,0,67,930.181406,945.0
